tools/discovery/discovery.go

```diff
@@ -0,0 +1,131 @@
+package discovery
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+	"github.com/xiekeyang/oci-discovery/tools/object"
+)
+
+var (
+	defaultTrans = &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+)
+
+func DiscoveryHandler(context *cli.Context) error {
+	var (
+		name  = context.Args()[0]
+		roots []v1.Descriptor
+	)
+
+	v, err := paramsParser(name)
+	if err != nil {
+		return err
+	}
+
+	engines, err := refEnginesFetching(v)
+	if err != nil {
+		return err
+	}
+
+	for _, engine := range engines.RefEngines {
+		var ur urlResolver = urlResolver(engine.Uri)
+		u, err := ur.resolve(v)
+		if err != nil {
+			return err
+		}
+
+		index, err := ociIndexFetching(u)
+		if err != nil {
+			return err
+		}
+
+		if fragment, ok := v["fragment"]; ok {
+			for _, manifest := range index.Manifests {
```

Comment:
This logic is specific to the [OCI Index Template Protocol][1], and belongs somewhere outside of the ref-engine-discovery implementation.

[1]: https://github.com/xiekeyang/oci-discovery/blob/44ec3cf3113e29a743ad04220ccb7ff5197dab2a/index-template.md